### ISSUE TITLE
Requiring a native module twice, fails

### ISSUE THREAD

**Issue by t-moe** [patriksimek/vm2#29]
When I try to load a certain native module twice I get a strange exception.

> VMError: Failed to load '<...>/deasync.node': [Module did not self-register.]

Here's a minimal example, which shows the problem. (Requires the npm module 'deasync' to run).
```
//Look up module path for this platform:
var path = require('path');
var fs = require('fs');

var nodeV = 'node-' + /[0-9]+\.[0-9]+/.exec(process.versions.node)[0];
var nodeVM = 'node-' + /[0-9]+/.exec(process.versions.node)[0];
var modPath = path.join(__dirname,'node_modules','deasync', 'bin', process.platform + '-' + process.arch + '-' + nodeV, 'deasync');

try{
    fs.statSync(modPath + '.node');
}
catch(ex){
    modPath = path.join(__dirname, 'node_modules','deasync','bin', process.platform + '-' + process.arch + '-' + nodeVM, 'deasync');
    fs.statSync(modPath + '.node');
}

//Beginning of the actual problem:

require(modPath); //Require native module for the first time

const NodeVM = require('vm2').NodeVM;
const vm = new NodeVM({
    require: true,
    requireExternal: true
});

vm.run("module.exports = require("+JSON.stringify(modPath)+")"); //Require module for the 2nd time, now inside vm.

//Exception occurs: VMError: Failed to load '<...>/deasync.node': [Module did not self-register.]
//Exception occurs as well if we require the same native module in two separate instances of NodeVm.
```

**Comment by Enteee** [comment]
an other version of the problem:

works:
```javascript
const NodeVM = require('vm2').NodeVM;
const vm = new NodeVM({
    require: true,
    requireExternal: true
});

vm.run("require('deasync')", __filename);
```

does not work:

```javascript
require('deasync');

const NodeVM = require('vm2').NodeVM;
const vm = new NodeVM({
    require: true,
    requireExternal: true
});

vm.run("require('deasync')", __filename);
```
Exception:
```
/home/ente/workspace/fluentflow/node_modules/bindings/bindings.js:85
        throw e
        ^
VMError: Module '/home/ente/workspace/fluentflow/node_modules/deasync/build/deasync.node' not found
    at _require (sandbox.js:138:17)
    at bindings (/home/ente/workspace/fluentflow/node_modules/bindings/bindings.js:78:44)
    at /home/ente/workspace/fluentflow/node_modules/deasync/index.js:24:31
    at _require (sandbox.js:190:9)
    at Object.<anonymous> (/home/ente/workspace/fluentflow/test.js:1:63)
    at NodeVM.run (/home/ente/workspace/fluentflow/node_modules/vm2/lib/main.js:393:13)
    at Object.<anonymous> (/home/ente/workspace/fluentflow/test.js:9:4)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
```

does not work:
```javascript
const NodeVM = require('vm2').NodeVM;
const vm = new NodeVM({
    require: true,
    requireExternal: true
});

vm.run("require('deasync')", __filename);

const vm2 = new NodeVM({
    require: true,
    requireExternal: true
});
vm2.run("require('deasync')", __filename);
```

Exception:
```
/home/ente/workspace/fluentflow/node_modules/bindings/bindings.js:85
        throw e
        ^
VMError: Module '/home/ente/workspace/fluentflow/node_modules/deasync/build/deasync.node' not found
    at _require (sandbox.js:138:17)
    at bindings (/home/ente/workspace/fluentflow/node_modules/bindings/bindings.js:78:44)
    at /home/ente/workspace/fluentflow/node_modules/deasync/index.js:24:31
    at _require (sandbox.js:190:9)
    at Object.<anonymous> (/home/ente/workspace/fluentflow/test.js:1:63)
    at NodeVM.run (/home/ente/workspace/fluentflow/node_modules/vm2/lib/main.js:393:13)
    at Object.<anonymous> (/home/ente/workspace/fluentflow/test.js:15:5)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
```

**Comment by patriksimek** [comment]
Thanks for the report, both failures are cause by `Module did not self-register.`. Since I have almost no experience with creating native modules, any help is very welcome.

**Comment by t-moe** [comment]
Ok. Here are some thoughts:
* `Module did not self register` is coming from [here in process.dlopen](https://github.com/nodejs/node/blob/ea63f79752e286c2987740a52b81cd289d125806/src/node.cc#L2302). 
* There's an [interesting comment ](https://github.com/nodejs/node/blob/ea63f79752e286c2987740a52b81cd289d125806/src/node.cc#L2265-L2267) above that function:
> // FIXME(bnoordhuis) Not multi-context ready. TBD how to resolve the conflict
> // when two contexts try to load the same shared object. Maybe have a shadow
> // cache that's a plain C list or hash table that's shared across contexts?

* I'm not familiar with the vm2 internals. Are we in a multi context scenario here?
* If so, then this is probably linked to https://github.com/nodejs/node/issues/5016 and we might have to ask them for help.

**Comment by patriksimek** [comment]
Yes, vm2 uses multi-context internally. This is definitely the core of the problem. It might be also somehow connected with #22 (loading a buffer in new context breaks buffer in original context).

**Comment by patriksimek** [comment]
Fixed in 3.0.
